consensus/consensus_v2.go

```diff
@@ -679,13 +681,75 @@ func (consensus *Consensus) commitBlock(blk *types.Block, committedMsg *FBFTMess
 	return nil
 }
 
+func (consensus *Consensus) rotateLeader(epoch *big.Int) {
+	prev := consensus.GetLeaderPubKey()
+	curNumber := consensus.Blockchain.CurrentHeader().Number().Uint64()
+	utils.Logger().Info().Msgf("[Rotating leader] epoch: %v rotation:%v numblocks:%d", epoch.Uint64(), consensus.Blockchain.Config().IsLeaderRotation(epoch), consensus.Blockchain.Config().LeaderRotationBlocksCount)
+	if consensus.Blockchain.Config().IsLeaderRotation(epoch) {
+		leader := consensus.GetLeaderPubKey()
+		for i := 0; i < consensus.Blockchain.Config().LeaderRotationBlocksCount; i++ {
+			header := consensus.Blockchain.GetHeaderByNumber(curNumber - uint64(i))
+			if header == nil {
+				return
+			}
+			// Previous epoch, we should not change leader.
+			if header.Epoch().Uint64() != epoch.Uint64() {
+				return
+			}
+			// Check if the same leader.
+			pub, err := chain.GetLeaderPubKeyFromCoinbase(consensus.Blockchain, header)
+			if err != nil {
+				utils.Logger().Error().Err(err).Msg("Failed to get leader public key from coinbase")
+				return
+			}
+			if !pub.Object.IsEqual(leader.Object) {
+				// Another leader.
+				return
+			}
+		}
+		// Passed all checks, we can change leader.
+		var (
+			wasFound bool
+			next     *bls.PublicKeyWrapper
+		)
+		// The same leader for N blocks.
+		if consensus.Blockchain.Config().IsAllowlistEpoch(epoch) {
```

Comment:
Let's remove the allowlist logic for now. We may not need to use it eventually.

Reply:
Removed. I can remove `IsAllowlistEpoch` method and logic if we don't need it anymore in another pr.